### PR TITLE
[Utils][SPIR-V] Adding spirv-sim to LLVM

### PR DESCRIPTION
**Pull request by Keenuts**
Currently, the testing infrastructure for SPIR-V is based on FileCheck. Those tests are great to check some level of codegen, but when the test needs check both the CFG layout and the content of each basic-block, things becomes messy.

- Because the CHECK/CHECK-DAG/CHECK-NEXT state is limited, it is sometimes hard to catch the good block: if 2 basic blocks have similar instructions, FileCheck can match the wrong one.

- Cross-lane interaction can be a bit difficult to understand, and writting a FileCheck test that is strong enough to catch bad CFG transforms while not being broken everytime some unrelated codegen part changes is hard.

And lastly, the spirv-val tooling we have checks that the generated SPIR-V respects the spec, not that it is correct in regards to the source IR.

For those reasons, I believe the best way to test the structurizer is to:
 - run spirv-val to make sure the CFG respects the spec.
 - simulate the function to validate result for each lane, making sure the generated code is correct.

This simulator has no other dependencies than core python. It also only supports a very limited set of instructions as we can test most features through control-flow and some basic cross-lane interactions.

As-is, the added tests are just a harness for the simulator itself. If this gets merged, the structurizer PR will benefit from this as I'll be able to add extensive testing using this.

### CODE REVIEW

File: llvm/utils/spirv-sim/instructions.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+class OpConstantComposite(OpConstant):
+    def static_execution(self, lane):
+        result = []
+        length = self.get_register(self._operands[0])
+        for op in self._operands[1:]:
+            result.append(self.get_register(op))
+        lane.set_register(self._result, result)
+
+
+class OpConstantComposite(OpConstant):
```
Comment (michalpaszkowski) [inline llvm/utils/spirv-sim/instructions.py:141]: `OpConstantComposite` is defined twice

File: llvm/utils/spirv-sim/spirv-sim.py [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+    def instruction(self):
+        return self.function[self.basic_block][self.instruction_index]
+
+    # Increment IP by 1. This only works inside a basic-block boundary.
+    # Incrementing IP when at the boundary of a basic block will fail.
+    def __add__(self, value: int):
+        bb = self.function[self.basic_block]
+        assert len(bb) > self.instruction_index + value
+        return InstructionPointer(
+            self.function, self.basic_block, self.instruction_index + 1
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/utils/spirv-sim/spirv-sim.py:172]: right, that's correct. Just + is never used with another value than 1 so haven't caught that.

File: llvm/utils/spirv-sim/spirv-sim.py [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        try:
+            with open(filename, "r") as f:
+                lines = f.read().split("\n")
+        except Exception:  # (FileNotFoundError, PermissionError):
+            return []
+    else:
+        lines = sys.stdin.readlines()
+
+    # Remove leading/trailing whitespaces.
+    lines = [x.rstrip().lstrip() for x in lines]
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/utils/spirv-sim/spirv-sim.py:569]: Ah that's better.

File: llvm/utils/spirv-sim/spirv-sim.py [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+        self._ip: InstructionPointer = None
+        # If this lane running.
+        self._running: bool = True
+        # The wave this lane belongs to.
+        self._wave: Wave = wave
+        # The callstack of this lane. Each tuple represents 1 call.
+        #   The first element is the IP the function will return to.
+        #   The second element is the callback to call to store the return value
+        #   into the correct register.
+        self._callstack: list[tuple(InstructionPointer, Callback)] = []
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/utils/spirv-sim/spirv-sim.py:191]: Ahh thanks, sounds right.

File: llvm/utils/spirv-sim/instructions.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
+# Base class for an instruction. To implement a basic instruction that doesn't
+# impact the control-flow, create a new class inheriting from this.
+class Instruction:
+    _result: str
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply llvm/utils/spirv-sim/instructions.py:4]: Looks like class members can be annotated either here or in the __init__. Moved all members annotation in the class body, and removed those from the constructor.

**Review by michalpaszkowski — APPROVED** [review]

**Review by iliya-diyachkov — COMMENTED** [review]
Please fix issues in comments.

**Comment by Keenuts** [comment]
Thanks! Typos fixed, & rebased

**Pull request merged**